File: leetcode/0515_find_largest_value_in_each_tree_row.cs
// https://leetcode.com/problems/find-largest-value-in-each-tree-row/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> LargestValues(TreeNode root) {
        List<int> result = new List<int>();
        Dictionary<int, int> maxs = new Dictionary<int, int>();

        Queue<(TreeNode, int)> q = new Queue<(TreeNode, int)>();
        q.Enqueue((root, 0));
        
        while (q.Count > 0)
        {
            (TreeNode t, int d) = q.Dequeue();
            if (t == null) continue;
            
            maxs[d] = Math.Max(maxs.TryGetValue(d, out int max) ? max : t.val, t.val);
            q.Enqueue((t.left, d + 1));
            q.Enqueue((t.right, d + 1));
        }
        
        return maxs.OrderBy(pair => pair.Key).Select(pair => pair.Value).ToList();
    }
}

File: leetcode/0515_find_largest_value_in_each_tree_row.cpp
// https://leetcode.com/problems/find-largest-value-in-each-tree-row/description

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    struct item
    {
        TreeNode *pt;
        int d;
    };

public:
    vector<int> largestValues(TreeNode* root) {
        map<int,int> maxs;
        queue<item> q;
        q.push({root, 0});
        
        while (!q.empty())
        {
            auto e = q.front(); q.pop();
            if (e.pt == nullptr) continue;
            
            auto p = maxs.find(e.d);
            maxs[e.d] = max(p == maxs.cend() ? numeric_limits<int>::min() : p->second, e.pt->val);
            q.push({e.pt->left, e.d + 1});
            q.push({e.pt->right, e.d + 1});
        }

        vector<int> result;
        result.reserve(maxs.size());
        for (auto&& e : maxs) result.push_back(e.second);
        return result;
    }
};
